crates/services/gas_price_service/src/v1/service.rs

```diff
@@ -32,94 +44,85 @@ use fuel_core_services::{
 {
     /// The algorithm that can be used in the next block
     shared_algo: SharedV1Algorithm,
     /// The L2 block source
     l2_block_source: L2,
-    /// The metadata storage
-    metadata_storage: Metadata,
     /// The algorithm updater
     algorithm_updater: AlgorithmUpdaterV1,
     /// the da source adapter handle
     da_source_adapter_handle: DaSourceService<DA>,
     /// The da source channel
     da_source_channel: Receiver<DaBlockCosts>,
+    /// Buffer of block costs from the DA chain
+    da_block_costs_buffer: Vec<DaBlockCosts>,
+    /// Storage transaction provider for metadata and unrecorded blocks
+    pub(crate) storage_tx_provider: StorageTxProvider,
 }
 
-impl<L2, Metadata, DA> GasPriceServiceV1<L2, Metadata, DA>
+impl<L2, DA, StorageTxProvider> GasPriceServiceV1<L2, DA, StorageTxProvider>
 where
     L2: L2BlockSource,
-    Metadata: MetadataStorage,
     DA: DaBlockCostsSource,
+    StorageTxProvider: TransactionableStorage,
 {
-    async fn process_l2_block_res(
-        &mut self,
+    async fn commit_block_data_to_algorithm<'a>(
+        &'a mut self,
         l2_block_res: GasPriceResult<BlockInfo>,
-    ) -> anyhow::Result<()> {
+    ) -> anyhow::Result<()>
+    where
+        StorageTxProvider::Transaction<'a>:
+            SetMetadataStorage + UnrecordedBlocks + SetDaSequenceNumber,
+    {
         tracing::info!("Received L2 block result: {:?}", l2_block_res);
         let block = l2_block_res?;
 
         tracing::debug!("Updating gas price algorithm");
         self.apply_block_info_to_gas_algorithm(block).await?;
         Ok(())
     }
-
-    async fn process_da_block_costs_res(
-        &mut self,
-        da_block_costs: Result<DaBlockCosts, RecvError>,
-    ) -> anyhow::Result<()> {
-        tracing::info!("Received DA block costs: {:?}", da_block_costs);
-        let da_block_costs = da_block_costs?;
-
-        tracing::debug!("Updating DA block costs");
-        self.apply_da_block_costs_to_gas_algorithm(da_block_costs)
-            .await?;
-        Ok(())
-    }
 }
 
-impl<L2, Metadata, DA> GasPriceServiceV1<L2, Metadata, DA>
+impl<L2, DA, StorageTxProvider> GasPriceServiceV1<L2, DA, StorageTxProvider>
 where
-    Metadata: MetadataStorage,
     DA: DaBlockCostsSource,
+    StorageTxProvider: TransactionableStorage,
 {
     pub fn new(
         l2_block_source: L2,
-        metadata_storage: Metadata,
         shared_algo: SharedV1Algorithm,
         algorithm_updater: AlgorithmUpdaterV1,
         da_source_adapter_handle: DaSourceService<DA>,
+        storage_tx_provider: StorageTxProvider,
     ) -> Self {
         let da_source_channel =
             da_source_adapter_handle.shared_data().clone().subscribe();
         Self {
             shared_algo,
             l2_block_source,
-            metadata_storage,
             algorithm_updater,
             da_source_adapter_handle,
             da_source_channel,
+            da_block_costs_buffer: Vec::new(),
```

Comment:
We only drain this buffer upon receiving L2 block (and on shutdown). I think there's no risk of unbounded growth then, but leaving the comment so you can double check :)

Reply:
Yeah. Should be good. If we aren't hitting L2 blocks to drain this, then we have a much bigger problem than memory growth.